### ISSUE TITLE
Feature request: dropout with noise_shape does not support varied batchsize

### ISSUE THREAD

**Issue by huzqatpku** [tensorflow/tensorflow#4987]
Since it's convenient to support varied batchsize in tf.placeholder by setting the head element of argument "shape" to None, tf.nn.dropout does not allow None in its argument "noise_shape", leading to models using dropout with noise_shape cumbersome to implement varied batchsize.  I wonder is it possible to allow None in noise_shape for better support for varied batchsize? Thank you!

**Comment by asimshankar** [comment]
@ebrevdo : Any thoughts?

**Comment by ebrevdo** [comment]
Noise shape may be a Tensor where any of the dimensions are not known at graph build time; i.e., you can have noise_shape = tf.Placeholder(dtype=tf.int32, shape=None)  ... or shape=[None, 3, 4])  ... or:

batch_size = tf.Placeholder(tf.int32, shape=())
noise_shape = tf.concat(0, [batch_size] + remaining_shape)

**Comment by ebrevdo** [comment]
or you can even do:

batch_size = tf.shape(input)[0]
noise_shape = tf.concat(0, [noise_shape] + remaining_shape)
